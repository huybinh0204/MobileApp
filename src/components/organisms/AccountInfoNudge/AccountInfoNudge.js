import { useNavigation } from '@react-navigation/native';
import React, { useEffect, useRef } from 'react';
import { TouchableOpacity } from 'react-native';
import BottomSheet from 'react-native-raw-bottom-sheet';
import { useDispatch, useSelector } from 'react-redux';
import { AccountNumber, IconSvg, MainButton } from '_components/atoms';
import { ListItem } from '_components/molecules';
import { COLORS, FEATURE_FLAGS, ICONS, NAVIGATION } from '_constants';
import { useBooleanFeatureFlag } from '_hooks';
import strings from '_localization';
import { AccountActions, AccountSelectors } from '_store/account';
import { CustomerSelectors } from '_store/customer';
import {
  FundDirectDepositStoreActions,
  FundDirectDepositStoreSelectors,
} from '_store/pageStore/fundDirectDepositStore';
import { normalize } from '_utilities/screen';
import {
  BottomSheetButton,
  ListItemDivider,
  MoveMoneyBottomSheetContainer,
  MoveMoneyBottomSheetTitle,
  NudgeCard,
  SecondaryText,
} from './AccountInfoNudge.styles';

const AccountInfoNudge = () => {
  const dispatch = useDispatch();
  const { navigate } = useNavigation();

  const externalId = useSelector(AccountSelectors.getExternalId);
  const customerExternalId = useSelector(CustomerSelectors.getExternalId);
  const isPhoneVerified = useSelector(CustomerSelectors.getVerificationStatus);
  const { accountNumber, routingNumber } = useSelector(
    FundDirectDepositStoreSelectors.getRenderData
  );

  const isAchInEnabled = useBooleanFeatureFlag(FEATURE_FLAGS.ACH_IN, customerExternalId);

  useEffect(() => {
    if (externalId && (!accountNumber || !routingNumber)) {
      dispatch(FundDirectDepositStoreActions.fetchRenderData(externalId));
    }
  }, [dispatch, accountNumber, routingNumber, externalId]);

  const modal = useRef(null);

  const goToTakeOutMoney = () => {
    modal.current?.close();

    if (isPhoneVerified) {
      navigate(NAVIGATION.accounts.takeOutMoney);
    } else {
      navigate(NAVIGATION.shared.phoneNumber, { target: NAVIGATION.accounts.takeOutMoney });
    }
  };

  const goToAddMoney = () => {
    modal.current?.close();

    if (isPhoneVerified) {
      navigate(NAVIGATION.shared.addMoney);
      dispatch(AccountActions.setLinkAchAccountSuccess(false));
    } else {
      navigate(NAVIGATION.shared.phoneNumber, { target: NAVIGATION.shared.addMoney });
    }
  };

  const goToKinlyAccount = () => {
    navigate(NAVIGATION.settings.accountInfo);
  };

  return (
    <>
      <NudgeCard>
        <AccountNumber onPressAccount={goToKinlyAccount} />
        <MainButton onPress={() => modal.current?.open()} testID="AccountInfoTransferFunds">
          {strings.transferMoney.transferFunds}
        </MainButton>
      </NudgeCard>
      <BottomSheet closeOnPressMask height={normalize(350)} ref={modal}>
        <MoveMoneyBottomSheetContainer>
          <MoveMoneyBottomSheetTitle>{strings.transferMoney.title}</MoveMoneyBottomSheetTitle>
          <SecondaryText disabled={!isAchInEnabled}>{strings.transferMoney.subTitle}</SecondaryText>
          <TouchableOpacity onPress={goToAddMoney} testID="AddMoney" disabled={!isAchInEnabled}>
            <ListItem
              icon={isAchInEnabled ? ICONS.finances : ICONS.financesInactive}
              title={strings.transferMoney.addMoney}
              color={isAchInEnabled ? COLORS.alpha500 : COLORS.beta500}
              subtitle={isAchInEnabled ? null : strings.transferMoney.addMoneyDisclaimer}
              disabled={!isAchInEnabled}
            >
              <IconSvg
                icon={isAchInEnabled ? ICONS.arrowRightActive : ICONS.arrowRightInactive}
                height={normalize(25)}
                width={normalize(25)}
              />
            </ListItem>
          </TouchableOpacity>
          <ListItemDivider />
          <TouchableOpacity onPress={goToTakeOutMoney} testID="TakeOutMoney">
            <ListItem icon={ICONS.atm} title={strings.transferMoney.takeOutMoney}>
              <IconSvg icon={ICONS.arrowRightActive} height={normalize(25)} width={normalize(25)} />
            </ListItem>
          </TouchableOpacity>
          <BottomSheetButton
            onPress={() => modal.current?.close()}
            testID="transferMoneyBottomSheetCancelButton"
          >
            {strings.transferMoney.cancel}
          </BottomSheetButton>
        </MoveMoneyBottomSheetContainer>
      </BottomSheet>
    </>
  );
};

export default AccountInfoNudge;
