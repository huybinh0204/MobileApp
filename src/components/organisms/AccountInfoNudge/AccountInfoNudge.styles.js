import styled from '@emotion/native';
import { BaseText, Card, MainButton } from '_components/atoms';

export const NudgeCard = styled(Card)(({ theme }) => ({
  marginBottom: theme.spacing.s,
}));

export const Row = styled.View(({ theme }) => ({
  flexDirection: 'row',
  marginBottom: theme.spacing.xs,
}));

export const Column = styled.View({
  flex: 1,
  alignItems: 'center',
  justifyContent: 'center',
});

export const Title = styled(BaseText)(({ theme }) => ({
  ...theme.typography.tinyContentRegular,
  color: theme.colors.beta500,
}));

export const Value = styled(BaseText)(({ theme }) => ({
  ...theme.typography.heading6,
  color: theme.colors.beta900,
}));

export const MoveMoneyBottomSheetContainer = styled.View(({ theme }) => ({
  flex: 1,
  padding: theme.spacing.m,
}));

export const MoveMoneyBottomSheetTitle = styled(BaseText)(({ theme }) => ({
  ...theme.typography.heading5,
  color: theme.colors.beta900,
  marginBottom: theme.spacing.xs,
}));

export const SecondaryText = styled(BaseText)(({ theme }) => ({
  ...theme.typography.mainContentRegular,
  color: theme.colors.beta400,
  marginBottom: theme.spacing.m,
}));

export const ListItemDivider = styled.View(({ theme }) => ({
  marginVertical: theme.spacing.s,
  borderBottomColor: theme.colors.beta50,
  borderBottomWidth: 1,
}));

export const BottomSheetButton = styled(MainButton)(({ theme }) => ({
  marginTop: theme.spacing.m,
}));
