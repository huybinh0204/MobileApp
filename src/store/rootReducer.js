import { combineReducers } from 'redux';
import actionWatchReducer from 'redux-action-watch/lib/reducer';
import { resettableReducer } from 'reduxsauce';
import { AccountReducer } from './account';
import { AuthenticationReducer } from './authentication';
import { CustomerReducer } from './customer';
import { FundDirectDepositStoreReducer } from './pageStore/fundDirectDepositStore';
import { RegisterReducer } from './register';
import { TransferReducer } from './transfer';

const resettable = resettableReducer('RESET');

export default combineReducers({
  account: resettable(AccountReducer),
  auth: resettable(AuthenticationReducer),
  customer: resettable(CustomerReducer),
  fundDirectDepositStoreReducer: resettable(FundDirectDepositStoreReducer),
  register: RegisterReducer,
  transfer: resettable(TransferReducer),
  watcher: actionWatchReducer, // Must be last
});
