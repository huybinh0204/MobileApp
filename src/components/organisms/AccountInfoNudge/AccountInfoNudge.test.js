import * as Split from '@splitsoftware/splitio-react';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { createStore } from 'redux';
import { NAVIGATION } from '_constants';
import strings from '_localization/index';
import rootReducer from '_store/rootReducer';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import { navigationMock } from '_test/mocks/navigation';
import AccountInfoNudge from './AccountInfoNudge';

const mockDispatch = jest.fn();
const mockNavigation = navigationMock(jest.fn);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => mockNavigation,
}));

describe('AccountInfoNudge', () => {
  let mockStore;

  beforeEach(() => {
    mockStore = createStore(rootReducer, {
      fundDirectDepositStoreReducer: {
        renderData: {
          accountNumber: '123456798',
          routingNumber: '000000000',
        },
      },
    });
  });

  it('should match the snapshot', () => {
    jest.spyOn(Split, 'useClient').mockImplementation(() => ({
      getTreatment(flag) {
        return 'on';
      },
    }));
    const { toJSON } = render(
      <ComponentWithProviders component={AccountInfoNudge} store={mockStore} />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should show account and routing numbers', async () => {
    const { getByText } = render(
      <ComponentWithProviders component={AccountInfoNudge} store={mockStore} />
    );

    const accountNumber = getByText('123 ••• •••');
    const routingNumber = getByText('000 000 000');

    await waitFor(() => {
      expect(accountNumber).toBeTruthy();
      expect(routingNumber).toBeTruthy();
    });
  });

  it('should open the modal when button is pressed', async () => {
    const { getByTestId, getByText } = render(
      <ComponentWithProviders component={AccountInfoNudge} store={mockStore} />
    );

    const transferFundsButton = getByTestId('AccountInfoTransferFunds');

    await waitFor(() => {
      fireEvent.press(transferFundsButton);
    });

    expect(getByText(strings.transferMoney.title)).toBeTruthy();
  });

  it("should navigate to Money Movement Page when the 'Take out money' button is pressed", async () => {
    const { getByTestId } = render(
      <ComponentWithProviders component={AccountInfoNudge} store={mockStore} />
    );

    const transferFundsButton = getByTestId('AccountInfoTransferFunds');

    fireEvent.press(transferFundsButton);

    const takeOutMoneyButton = getByTestId('TakeOutMoney');

    await waitFor(() => {
      fireEvent.press(takeOutMoneyButton);
    });

    expect(mockNavigation.navigate).toHaveBeenCalledWith(NAVIGATION.shared.phoneNumber, {
      target: NAVIGATION.accounts.takeOutMoney,
    });
  });
});
