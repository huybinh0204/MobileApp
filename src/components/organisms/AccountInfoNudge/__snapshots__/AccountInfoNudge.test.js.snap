// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountInfoNudge should match the snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignSelf": "stretch",
                              "backgroundColor": "#FFFFFF",
                              "borderColor": "#F2F3F4",
                              "borderRadius": 10,
                              "borderWidth": 0,
                              "padding": 29,
                              "shadowColor": "#362E60",
                              "shadowOffset": Object {
                                "height": 7,
                                "width": 0,
                              },
                              "shadowOpacity": 0.1,
                              "shadowRadius": 10,
                            },
                            Array [
                              Object {
                                "marginBottom": 29,
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "flexDirection": "row",
                                "marginBottom": 14.5,
                              },
                              undefined,
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "flex": 1,
                                  "justifyContent": "center",
                                },
                                undefined,
                              ]
                            }
                          >
                            <Text
                              allowFontScaling={false}
                              style={
                                Array [
                                  Object {
                                    "color": "#5F6A77",
                                    "fontFamily": "Poppins",
                                    "fontSize": 21.5,
                                    "lineHeight": 30.5,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              Routing
                            </Text>
                            <View
                              accessibilityState={
                                Object {
                                  "disabled": true,
                                }
                              }
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                              testID="routingButton"
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": "#1F1D25",
                                      "fontFamily": "Poppins-SemiBold",
                                      "fontSize": 29,
                                      "lineHeight": 40,
                                    },
                                    undefined,
                                  ]
                                }
                              >
                                000 000 000
                              </Text>
                            </View>
                          </View>
                          <View
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "flex": 1,
                                  "justifyContent": "center",
                                },
                                undefined,
                              ]
                            }
                          >
                            <Text
                              allowFontScaling={false}
                              style={
                                Array [
                                  Object {
                                    "color": "#5F6A77",
                                    "fontFamily": "Poppins",
                                    "fontSize": 21.5,
                                    "lineHeight": 30.5,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              Account
                            </Text>
                            <View
                              accessibilityState={
                                Object {
                                  "disabled": false,
                                }
                              }
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                              testID="accountButton"
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": "#1F1D25",
                                      "fontFamily": "Poppins-SemiBold",
                                      "fontSize": 29,
                                      "lineHeight": 40,
                                    },
                                    undefined,
                                  ]
                                }
                              >
                                123 ••• •••
                              </Text>
                            </View>
                          </View>
                        </View>
                        <View
                          accessibilityState={
                            Object {
                              "disabled": false,
                            }
                          }
                          accessible={true}
                          collapsable={false}
                          focusable={true}
                          onClick={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Object {
                              "alignSelf": "stretch",
                              "backgroundColor": "#3D3397",
                              "borderRadius": 10,
                              "opacity": 1,
                              "paddingHorizontal": 45.5,
                              "paddingVertical": 31,
                            }
                          }
                          testID="AccountInfoTransferFunds"
                        >
                          <Text
                            allowFontScaling={false}
                            disabled={false}
                            style={
                              Array [
                                Object {
                                  "color": "#FFFFFF",
                                  "fontFamily": "Poppins-SemiBold",
                                  "fontSize": 25.5,
                                  "lineHeight": 35.5,
                                  "textAlign": "center",
                                },
                                undefined,
                              ]
                            }
                          >
                            Transfer funds
                          </Text>
                        </View>
                      </View>
                      <Modal
                        animationType="none"
                        hardwareAccelerated={false}
                        onRequestClose={[Function]}
                        supportedOrientations={
                          Array [
                            "portrait",
                            "portrait-upside-down",
                            "landscape",
                            "landscape-left",
                            "landscape-right",
                          ]
                        }
                        transparent={true}
                        visible={false}
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
